src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,675 @@
+use std::convert::TryInto;
+use std::env;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::unix::ffi::OsStringExt;
+use std::ptr;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+#[cfg(feature = "wayland")]
+use std::fs::File;
+
+#[cfg(feature = "wayland")]
+use memmap2::MmapOptions;
+
+#[cfg(feature = "x11")]
+use x11_dl::xlib_xcb::xcb_connection_t;
+#[cfg(feature = "x11")]
+use xkbcommon_dl::x11::XKBCOMMON_X11_HANDLE as XKBXH;
+
+use xkbcommon_dl::{
+    self as ffi, xkb_state_component, XKBCOMMON_COMPOSE_HANDLE as XKBCH, XKBCOMMON_HANDLE as XKBH,
+};
+
+use crate::{
+    event::ElementState,
+    keyboard::{Key, KeyCode, KeyLocation},
+};
+
+// TODO: Wire this up without using a static `AtomicBool`.
+static RESET_DEAD_KEYS: AtomicBool = AtomicBool::new(false);
+
+#[inline]
+pub(crate) fn reset_dead_keys() {
+    RESET_DEAD_KEYS.store(true, Ordering::SeqCst);
+}
+
+pub(crate) struct KbState {
+    #[cfg(feature = "x11")]
+    xcb_connection: *mut xcb_connection_t,
```

Comment:
Nit: I'd prefer a higher level wrapper in order to avoid raw pointers and the unsafety that comes with it. But that can be done in a future PR.